src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngKey.EC.cs

```diff
@@ -0,0 +1,185 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Cryptography;
+using System.Diagnostics;
+using BCRYPT_ECC_PARAMETER_HEADER = Interop.BCrypt.BCRYPT_ECC_PARAMETER_HEADER;
+
+namespace System.Security.Cryptography
+{
+    public sealed partial class CngKey : IDisposable
+    {
+        /// <summary>
+        /// Does the key use elliptic curve cryptography
+        /// </summary>
+        /// <returns></returns>
+        internal bool IsEcc()
+        {
+            return (AlgorithmGroup == CngAlgorithmGroup.ECDiffieHellman ||
+                AlgorithmGroup == CngAlgorithmGroup.ECDsa);
+        }
+
+        /// <summary>
+        /// Does the key represent a named curve (Win10+)
+        /// </summary>
+        /// <returns></returns>
+        internal bool IsECNamedCurve()
+        {
+            return IsECNamedCurve(Algorithm.Algorithm);
+        }
+
+        internal static bool IsECNamedCurve(string algorithm)
+        {
+            return (algorithm == CngAlgorithm.ECDiffieHellman.Algorithm ||
+                algorithm == CngAlgorithm.ECDsa.Algorithm);
+        }
+
+        internal string GetCurveName()
+        {
+            if (IsECNamedCurve())
+            {
+                // Win10+
+                return _keyHandle.GetPropertyAsString(KeyPropertyName.ECCCurveName, CngPropertyOptions.None);
+            }
+
+            // Use pre-Win10 APIs
+            return GetECSpecificCurveName(); 
+        }
+
+        private string GetECSpecificCurveName()
+        {
+            string algorithm = Algorithm.Algorithm;
+
+            if (algorithm == CngAlgorithm.ECDiffieHellmanP256.Algorithm || 
+                algorithm == CngAlgorithm.ECDsaP256.Algorithm)
+            {
+                return "nistP256";
+            }
+
+            if (algorithm == CngAlgorithm.ECDiffieHellmanP384.Algorithm ||
+                algorithm == CngAlgorithm.ECDsaP384.Algorithm)
+            {
+                return "nistP384";
+            }
+
+            if (algorithm == CngAlgorithm.ECDiffieHellmanP521.Algorithm ||
+                algorithm == CngAlgorithm.ECDsaP521.Algorithm)
+            {
+                return "nistP521";
+            }
+
+            Debug.Assert(false);
+            throw new PlatformNotSupportedException(string.Format(SR.Cryptography_CurveNotSupported, algorithm));
+        }
+
+        /// <summary>
+        ///     Return a CngProperty representing a named curve.
+        /// </summary>
+        internal static CngProperty GetPropertyFromNamedCurve(ECCurve curve)
+        {
+            string curveName = curve.Oid.FriendlyName;
+            unsafe
+            {
+                byte[] curveNamebytes = System.Text.Encoding.Unicode.GetBytes(curveName + '\0');
+                unsafe
+                {
+                    fixed (byte* pValue = curveNamebytes)
```

Comment:
Good catch on pValue. Not sure how that code devolved into that.
Not really a perf issue on the concat as it's just called once per curve, but since you have the sample ready I'll use it. Thanks